examples/output_voltage_measurement/measurement.py

```diff
@@ -0,0 +1,237 @@
+        source_session_info = _get_session_info_for_pin(reservation.session_info, input_pin)
+        measure_session_info = _get_session_info_for_pin(reservation.session_info, output_pin)
+        # Creates NI-DCPower and NI-VISA DMM sessions.
+        with _nidcpower_helpers.create_session(
+            source_session_info, service_options.use_simulation, grpc_device_channel
+        ) as source_session, _visa_helpers.create_session(
+            measure_session_info.resource_name,
+            use_simulation=service_options.use_simulation,
+        ) as measure_session:
+            pending_cancellation = False
+            _add_cancel_callback(source_session, measurement_service, pending_cancellation)
+
+            assert isinstance(measure_session, pyvisa.resources.MessageBasedResource)
+
+            log_instrument_id(measure_session)
+
+            # When this measurement is called from outside of TestStand (session_exists == False),
+            # reset the instrument to a known state. In TestStand, ProcessSetup resets the
+            # instrument.
+            if not measure_session_info.session_exists:
+                reset_instrument(measure_session)
+
+            channels = source_session.channels[source_session_info.channel_list]
+
+            channels.source_mode = nidcpower.SourceMode.SINGLE_POINT
+            channels.output_function = nidcpower.OutputFunction.DC_VOLTAGE
+            channels.current_limit = current_limit
+            channels.voltage_level_range = voltage_level_range
+            channels.current_limit_range = current_limit_range
+            channels.source_delay = hightime.timedelta(seconds=source_delay)
+            channels.voltage_level = voltage_level
+
+            # Configure NI-VISA DMM
+            function_enum = FUNCTION_TO_VALUE[measurement_type]
+            resolution_value = RESOLUTION_DIGITS_TO_VALUE[str(resolution_digits)]
+            measure_session.write("CONF:%s %.g,%.g" % (function_enum, range, resolution_value))
+            check_instrument_error(measure_session)
+
+            with channels.initiate():
+                _wait_for_source_complete_event(measurement_service, channels, pending_cancellation)
+
+            response = measure_session.query("READ?")
+            check_instrument_error(measure_session)
+            measured_value = float(response)
+
+            source_session = None  # Don't abort after this point
+
+    logging.info("Completed measurement: measured_value=%g", measured_value)
+    return (measured_value,)
+
+
+def _get_session_info_for_pin(session_info, pin_name):
+    session_index = get_session_and_channel_for_pin(session_info, pin_name)[0]
+    return session_info[session_index]
+
+
+def _add_cancel_callback(session, measurement_service, pending_cancellation):
+    def cancel_callback():
+        session_to_abort = session
+        if session_to_abort is not None:
+            nonlocal pending_cancellation
+            pending_cancellation = True
```

Comment:
This doesn't work due to Python scoping rules. Unfortunately, this minor refactor seems to have opened another "can of worms" that may lead to changing all of the examples.

The nested function `cancel_callback` can modify the `pending_cancellation` variable captured from the `_add_cancel_callback` function. However, this has no effect on the `measure` function because `bool` in Python behaves as if it is passed by value.

Really, it's more complicated: there are two singleton instances of `bool`: `True` and `False`, and you pass around references to them. However, these instances are immutable, so assigning to a `bool` variable works by rebinding it to a different reference. `_add_cancel_callback` cannot modify `measure`'s `pending_cancellation` variable.

For more details and explanation, read:
- https://realpython.com/python-scope-legb-rule/
- https://stackoverflow.com/questions/1494696/are-pythons-bools-passed-by-value

An alternative is to use `threading.Event` like in https://github.com/ni/measurementlink-python/blob/main/examples/ui_progress_updates/measurement.py . Event objects are mutable, so you can have a callback set the event without rebinding any references. As a plus, you don't even need to define a callback function, because you can use the event's `set` method as the callback.

In addition, calling `abort()` on the NI-DCPower session doesn't make as much sense as we originally thought. It doesn't unblock waiting API calls like it does in NI-DAQmx. Also, raising an exception in the `with session.initiate()` block automatically calls `abort()`, so we don't really need to do it again.

Reply:
I have updated the `measurement.py` to use `threading.Event` for "cancellation_callback".